File: packages/js-dpp/lib/stateTransition/fees/operations/ReadOperation.js
const AbstractOperation = require('./AbstractOperation');

class ReadOperation extends AbstractOperation {
  /**
   * @param {number} valueSize
   */
  constructor(valueSize) {
    super();

    this.valueSize = valueSize;
  }

  /**
   * Get CPU cost of the operation
   *
   * @returns {number}
   */
  getProcessingCost() {
    return ReadOperation.BASE_PROCESSING_COST + this.valueSize * ReadOperation.CREDIT_PER_BYTE;
  }

  /**
   * Get storage cost of the operation
   *
   * @returns {number}
   */
  getStorageCost() {
    return 0;
  }

  /**
   * Get operation type
   *
   * @returns {string}
   */
  getType() {
    return AbstractOperation.TYPES.READ;
  }
}

ReadOperation.CREDIT_PER_BYTE = 12;
ReadOperation.BASE_PROCESSING_COST = 8400;

module.exports = ReadOperation;

File: packages/rs-dpp/src/state_transition/fees/operations/read_operation.rs
use super::{OperationLike, OperationType};

pub const BASE_PROCESSING_COST: i64 = 8400;
pub const CREDIT_PER_BYTE: i64 = 12;

#[derive(Default, Debug, Clone, Copy)]
pub struct ReadOperation {
    value_size: usize,
}

impl ReadOperation {
    pub fn new(value_size: usize) -> Self {
        Self { value_size }
    }
}

impl OperationLike for ReadOperation {
    fn get_processing_cost(&self) -> i64 {
        BASE_PROCESSING_COST + (self.value_size as i64 * CREDIT_PER_BYTE)
    }

    fn get_storage_cost(&self) -> i64 {
        0
    }

    fn get_type(&self) -> OperationType {
        OperationType::Read
    }
}
